spec/requests/works_show_request_spec.rb

```diff
@@ -0,0 +1,80 @@
+require "spec_helper"
+
+describe "Works#show" do
+  let(:work) { create(:work) }
+  let(:chapter) { work.chapters.first }
+
+  context "when the first chapter of a work is unposted" do
+    before do
+      work.chapters.create(position: 1, posted: false, content: "Draft content")
+      chapter.update(position: 2)
+    end
+
+    context "when logged in as an admin" do
+      before { sign_in create(:admin), scope: :admin }
+
+      it "displays the unposted chapter" do
+        get work_url(work)
+        expect(response).to render_template(:show)
+        expect(response.body).not_to include(chapter.content)
+        expect(response.body).to include("Draft content")
+      end
+    end
+
+    context "when logged in as a random user" do
+      before do
+        allow_any_instance_of(ApplicationController).to receive(:logout_if_not_user_credentials).and_return(false)
+        sign_in create(:user), scope: :user
```

Comment:
I didn't use the existing `LoginMacros` because including it resulted in:

```
     Failure/Error: @request.env["devise.mapping"] = Devise.mappings[:admin]
     
     NoMethodError:
       undefined method `env' for nil:NilClass
       Did you mean?  end
     # ./spec/support/login_macros.rb:18:in `fake_login_admin'
     # ./spec/requests/works_show_request_spec.rb:17:in `block (4 levels) in <top (required)>'
```

Reply:
Are the lines with `@request.env["devise.mapping"]` still necessary? I just tried removing them, and all of the tests in `spec/controllers/comments_controller_spec.rb` (which has by far the most calls to `fake_login`, `fake_login_known_user`, and `fake_login_admin`) seemed to pass.

If they were removed, you could probably use `LoginMacros` here.

Reply:
It sounds like it *could* be necessary in some controllers, but we'll see!